### ISSUE TITLE
AES-CTR (and probably more stream ciphers) are slow when used via the FFI's Stream_Cipher mode.

### ISSUE THREAD

**Issue by aahajj** [randombit/botan#3925]
Hello everyone,

I have created a benchmarking tool to compare performance between libraries and have noticed a significant drop in performance with Botan's AES CTR. In my benchmark, I generated plaintexts with a length of 4096 Bytes for each round and measured the time it takes for 1 million iterations. For my performance test I have used the Botan C interface ffi.h.

Results:

AES-NI and SSSE3 enabled:

| Library   | Algorithm    | Time (seconds) |
|-----------|--------------|----------------|
|Botan 3.2.0|     	 AES-128     	 |0.554971       
|Botan 3.2.0 |   	 AES-128/CTR 	|77.982720
|Botan 3.2.0  |   	 AES-128/CBC 	|5.963908
|Botan 3.2.0   |  	 AES-128/CFB 	|7.617746

AES-NI disabled:

| Library   | Algorithm    | Time (seconds) |
|-----------|--------------|----------------|
| Botan 3.2.0 | AES-128      | 9.187737       |
| Botan 3.2.0 | AES-128/CTR  | 81.234424      |
| Botan 3.2.0 | AES-128/CBC  | 19.137964      |
| Botan 3.2.0 | AES-128/CFB  | 20.943194      |

 With AES-NI and SSSE3 disabled:

| Library    | Algorithm    | Time (seconds) |
|------------|--------------|----------------|
| Botan 3.2.0 | AES-128      | 66.311669     |
| Botan 3.2.0 | AES-128/CTR  | 135.916029    |
| Botan 3.2.0 | AES-128/CBC  | 153.400266    |
| Botan 3.2.0 | AES-128/CFB  | 151.787276    |

### Environment:
- **Operating System:** Ubuntu 22.04.3 LTS (Codename: jammy)
- **CPU:** AMD Ryzen 5 5600U with Radeon Graphics
  - **Architecture:** x86_64
  - **CPU(s):** 12
  - **Thread(s) per core:** 2
  - **Core(s) per socket:** 6
  - **CPU max MHz:** 4289,0000
  - **CPU min MHz:** 400,0000
 ```bash
lscpu | grep "Byte Order"
Byte Order:                         Little Endian
```
I suspect that the AES CTR encryption process is not being parallelized, which could be causing the significant drop in performance. However, I'm uncertain about the exact reason for this issue and would appreciate your assistance in investigating further.

Thank you and best regards,
Ahmed 

Edit 1: add Byte Order
Edit 2: fix a comma typo

**Comment by reneme** [comment]
Mhh, current working hypothesis: CTR-mode being a stream cipher, it features an `update_granularity()` of 1 byte (!). Other cipher modes typically use the underlying block size of the bare symmetric algorithm. Also, CTR-mode doesn't provide any authentication tags or paddings, hence, its `minimum_final_size()` is 0 bytes.

The FFI adapter layer uses this information to determine the update granularity it will use internally:

https://github.com/randombit/botan/blob/b84acd926d1ccc78524c69aee5d53279b9a5e361/src/lib/ffi/ffi_cipher.cpp#L48-L51

... and then slices the input data accordingly:

https://github.com/randombit/botan/blob/b84acd926d1ccc78524c69aee5d53279b9a5e361/src/lib/ffi/ffi_cipher.cpp#L187-L205

Therefore, we might actually pass the payload data byte-by-byte to the encryption/decryption routine. Obviously, that's awfully slow. 😨

Thanks again, for reporting this!

**Comment by aahajj** [comment]
Ah, I stumbled upon that function today too, and I was a bit puzzled by the update granularity being just 1 byte. But now that you mention it, it actually makes perfect sense! Thank you for the clarification! 

Oh, and by the way, AES-128/CCM also has an update granularity of 1 Byte. CCM, among the three AEAD algorithms I've tested, seems to be the slowest, and I think it might be due to the same issue.

**Comment by aahajj** [comment]
I couldn't help but notice a naming inconsistency between the function botan_cipher_get_ideal_granularity() in the  [website's](https://botan.randombit.net/handbook/api_ref/ffi.html#_CPPv434botan_cipher_get_ideal_granularity14botan_cipher_tP6size_t) documentation and its declaration in https://github.com/randombit/botan/blob/b84acd926d1ccc78524c69aee5d53279b9a5e361/src/lib/ffi/ffi.h#L561

**Comment by reneme** [comment]
For the record: [This pull request addresses the performance issues](https://github.com/randombit/botan/pull/3951) described in this ticket. While working on the performance fix, another (functional) issue was found in the same code location (https://github.com/randombit/botan/pull/3971). As a result, the performance fix didn't make it to the just-released 3.4.0 release. The mentioned bugfix landed and got released, however.

@aahajj if you have time, please repeat your measurements after applying the patch suggested in #3951. It should speed up the stream-cipher-esque modes quite significantly.

**Comment by aahajj** [comment]
Sorry for the late response. I've changed jobs and am officially no longer working on that project. I was about to redo the measurements again but found that the patch is not merged yet, pending review from @randombit. By the way, I obtained approval to publish the tool on my GitHub page if you want to take a look at it: [CBOS](https://github.com/aahajj/CBOS). Once the pull request is merged, I will happily do the measurements again.
